File: Coders/DSA Concepts/quick_sort.cpp
#include <bits/stdc++.h>
using namespace std;

int partition(int *arr,int s,int e){
    int pivot=arr[s];
    int cnt=0;

    for (int i = s+1; i <= e; i++)
    {
        if(arr[i]<=pivot){
            cnt++;
        }
    }

    int pivotIndex=s+cnt;
    swap(arr[pivotIndex],arr[s]);


    //partition
    int i=0,j=e;
    while (i<=pivotIndex && j>pivotIndex)
    {
        while (arr[pivotIndex]>=arr[i])i++;
        while(arr[pivotIndex]<arr[j])j--;

        if(i<pivotIndex && j>pivotIndex)swap(arr[i++],arr[j--]);
    }
    return pivotIndex;
    

    
}

void quickSort(int *arr,int s,int e){
    // s is starting index and e is ending index
    if(s>=e)return;

    int p = partition(arr, s, e);

    quickSort(arr, s, p-1);
    quickSort(arr, p+1, e);
}

int main(){
    int n;
    int arr[1000];
    cin>>n;
    for(int i=0;i<n;i++)
        cin>>arr[i];
    quickSort(arr,0,n-1);

    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }

    return 0; 
}

File: Coders/DSA Concepts/QuickSort.java
public class QuickSort {
	public static void main(String[] args) {
	    int arr[] = {1, 0};
	    quickSort(arr, 0, arr.length);
	    for(int i = 0; i< arr.length; i++){
	    	    System.out.println(arr[i]);
	    	
	    	}
	}
	public static void quickSort(int input [], int start, int end){
		    if(end - start < 2) {
		    	return;
		    	
		    }
		    int pivotIndex = partition(input, start, end);
		    quickSort(input, start, pivotIndex);
		    quickSort(input, pivotIndex + 1, end);
		
		
	}
	public static int partition(int input[], int start, int end){
		    int i = start;
		    int j = end;
		    int pivot = input[start];
		    
		    while(i<j){
		        while(i<j && pivot <= input[--j]);
		        if (i<j){
		            input[i] = input[j];		
		        }
		        while(i<j && pivot >= input[++i]);
		        if(i<j){
		            input[j] = input[i];	
		        }		
		    }
		    
		    input[j] = pivot;
		    return j;
	}
}
